From bf7e8348d0594026abf9b17356a935b31bf07274 Mon Sep 17 00:00:00 2001
From: Eric Blake <eblake@redhat.com>
Date: Mon, 10 Jul 2017 17:52:18 +0200
Subject: [PATCH] qemu-nbd: Ignore SIGPIPE

RH-Author: Eric Blake <eblake@redhat.com>
Message-id: <20170710175218.13682-1-eblake@redhat.com>
Patchwork-id: 75718
O-Subject: [RHEL-7.4.z qemu-kvm PATCH] qemu-nbd: Ignore SIGPIPE
Bugzilla: 1468107
RH-Acked-by: John Snow <jsnow@redhat.com>
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: Max Reitz <mreitz@redhat.com>

qemu proper has done so for 13 years
(8a7ddc38a60648257dc0645ab4a05b33d6040063), qemu-img and qemu-io have
done so for four years (526eda14a68d5b3596be715505289b541288ef2a).
Ignoring this signal is especially important in qemu-nbd because
otherwise a client can easily take down the qemu-nbd server by dropping
the connection when the server wants to send something, for example:

$ qemu-nbd -x foo -f raw -t null-co:// &
[1] 12726
$ qemu-io -c quit nbd://localhost/bar
can't open device nbd://localhost/bar: No export with name 'bar' available
[1]  + 12726 broken pipe  qemu-nbd -x foo -f raw -t null-co://

In this case, the client sends an NBD_OPT_ABORT and closes the
connection (because it is not required to wait for a reply), but the
server replies with an NBD_REP_ACK (because it is required to reply).

Signed-off-by: Max Reitz <mreitz@redhat.com>
Message-Id: <20170611123714.31292-1-mreitz@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 041e32b8d9d076980b4e35317c0339e57ab888f1)
Signed-off-by: Miroslav Rezanina <mrezanin@redhat.com>

Conflicts:
	qemu-nbd.c - context

Fixes CVE-2017-10664
Signed-off-by: Eric Blake <eblake@redhat.com>
---
 qemu-nbd.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/qemu-nbd.c b/qemu-nbd.c
index e0f4517..fc768eb 100644
--- a/qemu-nbd.c
+++ b/qemu-nbd.c
@@ -363,6 +363,11 @@ int main(int argc, char **argv)
     memset(&sa_sigterm, 0, sizeof(sa_sigterm));
     sa_sigterm.sa_handler = termsig_handler;
     sigaction(SIGTERM, &sa_sigterm, NULL);
+
+#ifdef CONFIG_POSIX
+    signal(SIGPIPE, SIG_IGN);
+#endif
+
     qemu_init_exec_dir(argv[0]);
 
     while ((ch = getopt_long(argc, argv, sopt, lopt, &opt_ind)) != -1) {
-- 
1.8.3.1

